components/collector/src/source_collectors/testng/tests.py

```diff
@@ -1,23 +1,46 @@
 """TestNG tests collector."""
 
 from typing import cast
+from xml.etree.ElementTree import Element  # nosec, Element is not available from defusedxml, but only used as type
 
 from base_collectors import XMLFileSourceCollector
 from collector_utilities.functions import parse_source_response_xml
-from model import SourceMeasurement, SourceResponses
+from model import Entities, Entity, SourceMeasurement, SourceResponses
 
 
 class TestNGTests(XMLFileSourceCollector):
     """Collector for TestNG tests."""
 
+    TEST_RESULT = dict(FAIL="failed", PASS="passed", SKIP="skipped")  # nosec, TestNG test status to test result mapping
+
     async def _parse_source_responses(self, responses: SourceResponses) -> SourceMeasurement:
         """Override to parse the tests for the TestNG XML."""
-        test_statuses_to_count = cast(list[str], self._parameter("test_result"))
+        test_results_to_count = [status.lower() for status in cast(list[str], self._parameter("test_result"))]
         test_count = 0
         total = 0
+        entities = Entities()
         for response in responses:
             tree = await parse_source_response_xml(response)
-            for test_status in test_statuses_to_count:
-                test_count += int(tree.get(test_status) or "0")
+            for test_result in test_results_to_count:
+                test_count += int(tree.get(test_result) or "0")
             total += int(tree.get("total") or "0")
-        return SourceMeasurement(value=str(test_count), total=str(total))
+            entities.extend(self.__entities(tree, test_results_to_count))
+        return SourceMeasurement(value=str(test_count), total=str(total), entities=entities)
+
+    @classmethod
+    def __entities(cls, tree: Element, test_results_to_count: list[str]) -> Entities:
+        """Transform the test methods into entities."""
+        entities = Entities()
+        for test_class in tree.findall(".//class"):
+            class_name = test_class.get("name", "")
+            for test_method in test_class.findall(".//test-method"):
+                test_result = cls.TEST_RESULT.get(test_method.get("status", ""))
+                if test_method.get("is-config", "false") == "true" or test_result not in test_results_to_count:
+                    continue  # Skip config (beforeClass/afterClass) methods and test results the user wants to ignore
+                name = test_method.get("name", "")
+                description = test_method.get("description", "")
+                key = f"{class_name}_{name}"
```

Comment:
Does this key have to be unique? because this could theoretically lead to a key "_" for all entities without class_name and name. In the end I'm wondering if default to empty string makes sense for class_name and name.

Reply:
It does have to be unique. So far from looking a testng-result.xml files it seems test class name and test method name should always be present and the combination should be unique. Unfortunately, there's no DTD for testng-result.xml, so I'm not 100% sure. I'll add some comments.